src/pushed_authorization.rs

```diff
@@ -0,0 +1,269 @@
+        audience,
+        issuer: client_id,
+        nonce: None,
+        controller: ProofOfPossessionController { vm: None, jwk },
+    };
+    let pop = ProofOfPossession::generate(&pop_params, Duration::minutes(10));
+    let jwt = pop.to_jwt()?;
+
+    Ok(jwt)
+}
+
+pub async fn decode_jwt(jwt: String) -> Result<ProofOfPossessionBody, ParsingError> {
+    use ssi::{
+        jwk::Algorithm,
+        jws::{self, Header},
+        jwt,
+    };
+
+    let header: Header = jws::decode_unverified(jwt.as_str())?.0;
+
+    if header.type_ != Some(JWS_TYPE.to_string()) {
+        return Err(ParsingError::InvalidJWSType {
+            actual: format!("{:?}", header.type_),
+            expected: JWS_TYPE.to_string(),
+        });
+    }
+    if header.algorithm == Algorithm::None {
+        return Err(ParsingError::MissingJWSAlg);
+    }
+    let jwk = match header.jwk {
+        Some(jwk) => jwk,
+        None => return Err(ParsingError::MissingKeyParameters),
+    };
+    let decoded_jwt: ProofOfPossessionBody = jwt::decode_verify(jwt.as_str(), &jwk)?;
+    Ok(decoded_jwt)
+}
+
+pub async fn verify(
+    jwt: String,
+    client_id: String,
+    audience: url::Url,
+) -> Result<(), VerificationError> {
+    use crate::proof_of_possession::{ProofOfPossession, ProofOfPossessionVerificationParams};
+    use did_jwk::DIDJWK;
+
+    let pop = ProofOfPossession::from_jwt(jwt.as_str(), &DIDJWK)
+        .await
+        .unwrap();
+
+    pop.verify(&ProofOfPossessionVerificationParams {
+        nonce: pop.body.nonce.clone(),
+        audience,
+        issuer: client_id,
+        controller_did: None,
+        controller_jwk: None,
+        nbf_tolerance: None,
+        exp_tolerance: None,
+    })
+    .await
+}
+
+pub struct PushedAuthorizationRequest<'a, AD>
+where
+    AD: AuthorizationDetaislProfile,
+{
+    inner: oauth2::AuthorizationRequest<'a>, // TODO
+    par_auth_url: AuthUrl,
+    authorization_details: Vec<AuthorizationDetail<AD>>,
+    wallet_issuer: Option<IssuerUrl>, // TODO SIOP related
+    user_hint: Option<String>,
+    issuer_state: Option<CsrfToken>,
+}
+
+impl<'a, AD> PushedAuthorizationRequest<'a, AD>
+where
+    AD: AuthorizationDetaislProfile,
+{
+    pub(crate) fn new(
+        inner: oauth2::AuthorizationRequest<'a>,
+        par_auth_url: AuthUrl,
+        authorization_details: Vec<AuthorizationDetail<AD>>,
+        wallet_issuer: Option<IssuerUrl>,
+        user_hint: Option<String>,
+        issuer_state: Option<CsrfToken>,
+    ) -> Self {
+        Self {
+            inner,
+            par_auth_url,
+            authorization_details,
+            wallet_issuer,
+            user_hint,
+            issuer_state,
+        }
+    }
+
+    pub fn request(
+        self,
+        client_assertion_type: Option<String>,
+        client_assertion: Option<String>,
+    ) -> Result<(String, serde_json::Value, CsrfToken), serde_json::Error> {
+        let (url, token) = self.inner.url();
+        let mut body = json!({});
```

Comment:
I'm a bit confused by this function, I think what would be better is:
1. deserialise the query params using `serde_urlencoded`
2. define a type for the body to avoid building the json value directly